### ISSUE TITLE
Include SDE build steps in the dockerfile itself & do some cleanup

### ISSUE THREAD

**Issue by chrispsommers** [ipdk-io/ipdk#11]
The steps starting at https://github.com/ipdk-io/ipdk/blob/5e737af4c302013d5a50a5fd287ed43ec8366ef4/build/IPDK_Container/README#L53 are done from within the container. This adds manual steps and also means the changes are ephemeral - if you lose the container you'd have to redo everything, and it is fairly time-consuming. (It took nearly a *full day* to completely build and modify the container - I have slowish downlink)  In my case I did a docker commit of the running container to a new image to preserve it.

Could those steps 4-5, (possibly #6) be moved into the dockerfile to reduce effort and result in a more complete, durable image?

Also, I think the image could benefit from some cleanup after all the installing is done. The `docker commit` following SDE install produced a 13GB image! See below - the first image is the committed container after SDE install.
```
$ docker images
REPOSITORY         TAG       IMAGE ID       CREATED       SIZE
ipdk               latest    b801f17e5efd   11 days ago   13.3GB
ipdk/p4-ovs-fc33   1e7b540   96137ba4f7bf   11 days ago   1.67GB
fedora             33        a5465556eeb2   6 weeks ago   172MB
```

**Comment by stolsma** [comment]
I agree that this container build process and resulting container isn't as optimal as it should be for other then hard-core development. A lot need to be done to streamline the creation of (a) container(s) without all the build artefacts and libraries that are not needed any more when doing testing (or later in production). For me the total build process is 35 minutes and not really a problem (at the moment), but the resulting ultra large image is a pain in the ass... ;-)

**Comment by vsureshkumarp** [comment]
I Agree. With current Container approach, need to repeat the entire process again when we loose the container. Build process consumes considerable amount of time based on the type of server environment used for build procedure. 
Definitely couple of optimizations needed for reducing the Docker image size. 

Soon we are gong to commit the following changes to reduce the Docker size and address the dependent libraries, source code persistence with base container image.

Docker file modified to
1. Clone all required repositories, build and install all required modules.
2. Remove all dependent module source code.
3. Remove all build artifacts in source directories once after installing the modules either in standard path or custom user specified path.
4. Remove redundant standard packages

There is some room to optimize the amount of memory using for standard packages too.
We will be addressing this in next immediate commits.

Another major contributor for Docker image size is source code. We are working on approach which helps to reduce the image size without loosing the flexibility for development environment.
Idea is to install all the libraries and binaries required to run the scenarios by default and provide utility scripts to pull the source code on demand basics.

**Comment by stolsma** [comment]
Sounds good... I will wait for the commits then before working on them myself!

**Comment by chrispsommers** [comment]
Suggest to provide a build option (e.g. env variable) to retain the source, so a subsequent step isn't required to pull it in. Pulling it in afterwards is not a bad idea either.

**Comment by jfingerh** [comment]
I have been experimenting this weekend with variants of the scripts that do all cloning and compiling in the base Linux system itself, with no Docker involved.  I have nothing against Docker, other than my current ignorance in debugging issues specific to that :-)   You can't get any more persistent than storing changes in the Linux system's base file system, though.

**Comment by mestery** [comment]
Note that I've taken a step in this direction with #25 and will merge that today, we should build off of this in terms of the Docker images. I may take a pass at slimming things down a bit.

**Comment by stolsma** [comment]
I've build the a new docker image after #29 was pulled to main and it build to a 'nice' 6.63 GByte. A lot better then old process steps :-) If I find some obvious 'low hanging fruit' to slim it down further I will report it...

**Comment by vsureshkumarp** [comment]
With latest commits,
1. The complete build procedure is part of the docker build procedure.
2. Cleanup logic added to remove build artifacts once module libraries and binaries are installed.
3. To reduce the image size, added an additional user configurable option which indicates whether to keep all module source code or not. By default, source code is removed once after installing all the binaries.

With latest improvements, overall image size is 
             < 5.6 GB when source code is part of the image
             < 3.5 GB when source code is not part of the image.

**Comment by chrispsommers** [comment]
Excellent progress, thanks!

**Comment by vsureshkumarp** [comment]
@chrispsommers, Can I close this issue as most of the key image size contributors optimization have already taken care?

**Comment by chrispsommers** [comment]
Sure, thanks for following up. I wasn't sure about the "close issue" protocol on this project. Be my guest!

**Comment by vsureshkumarp** [comment]
@chrispsommers, Thanks. Closing issue.
